maintainers/scripts/haskell: unify pkg set update into single commit

sternenseemann opened the pull request:
The package set update should be a single commit, so metadata/configuration and hackage-packages.nix is always consistent. While implementing this, I've made some smaller cleanups and added documentation, see the commits for details. The last commit in the PR is the result of running the new script.

- [ ] I want to make the commit message of `mark-broken.sh` wrap at <80 characters as well which I'll do in a separate PR.
- [x] I also would like to remove the `--do-commit` flag from `update-stackage.sh` and `update-hackage.sh` where it no longer makes sense. #425204

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

sternenseemann committed:
maintainers/haskell/update*: print version diff when not committing

This allows other scripts to detect whether anything changed without
resorting to git-diff(1): If nothing changed, stdout will be empty (i.e.
we now enforce that other messages go to stderr). To make sure that this
behavior is retained in the future, the scripts' behavior is briefly
documented in the files' header.

sternenseemann committed:
maintainers/scripts/haskell: briefer commit message bodies

The “This commit was generated by …” message is very wordy and often
exceeds 72 characters which is a good (loose) target for wrapping lines
in commit messages.

sternenseemann committed:
maintainers/haskell/regen-h-packages.sh: let user edit commit msg

Calling git commit with --edit here, allows the user to rephrase the
very nondescript default message, in doing so e.g. clarifying why the
regeneration was necessary etc. This should hopefully encourage better
commit mesages.

sternenseemann committed:
maintainers/scripts/haskell: unify pkg set update into single commit

Currently, every package set consists of three commits, generated by
update-hackage.sh, update-stackage.sh and
regenerate-hackage-packages.sh, respectively. This is suboptimal, as it
necessarly causes intermediate states of Nixpkgs where the generated
hackage-packages.nix and all-cabal-hasehs and/or the hackage2nix
configuration files are out of sync. Ideally, running
regenerate-hackage-packages.sh is a no-op for every Nixpkgs revision.

This is achieved by adding a wrapper script, update-package-set.sh,
which runs the individual moving parts and commits the result.

sternenseemann commented on maintainers/scripts/haskell/update-package-set.sh:
```diff
@@ -0,0 +1,53 @@
+  exit 100
+fi
+
+stackage_diff="$(./maintainers/scripts/haskell/update-stackage.sh)"
+hackage_diff="$(./maintainers/scripts/haskell/update-hackage.sh)"
```
Makes sense, but all scripts currently assume this, so I'd do this in a follow up change at some point.

sternenseemann commented on maintainers/scripts/haskell/merge-and-open-pr.sh:
Seems like a good idea, though I'm not sure that it is really worth the CI time and hassle maintaining the workflow given that only very few PR touch the relevant files.

maralorn commented on maintainers/scripts/haskell/update-package-set.sh:
```diff
@@ -0,0 +1,53 @@
+if [[ -n "$stackage_diff" ]]; then
+  commit_message="haskellPackages: stackage $stackage_diff"
+  if [[ -n "$hackage_diff" ]]; then
+    commit_message="$commit_message"'
+
+all-cabal-hashes: '"$hackage_diff"
```
Oh, boy. Took me a while to grok this syntax … 😆

maralorn reviewed (approved):
LGTM, nice.

sternenseemann committed:
haskellPackages: stackage LTS 23.24 -> LTS 23.27

all-cabal-hashes: 2025-06-01T18:10:16Z -> 2025-07-07T21:33:55Z

(generated by maintainers/scripts/haskell/update-package-set.sh)
